### PR TITLE
DAOS-2809 dtx: synchronously commit DTX that creates dkey

### PR DESCRIPTION
**Pull request by Nasf-Fan**
Dkey is the determining factor for on which target the data will
be stored: the akeys and SV/EV tree records belong to the same
dkey will be stored on the same target. But for different dkeys
belong to the same redundancy group (for the same replicated object),
the leaders replica for these different dkeys may reside on different
targets in the same redundancy group.

For VOS_ITER_DKEY operation, if some new dkeys which DTXs are
committable but related leader replicas are on different targets,
then in spite of list dkeys on which replica, the client will get
-DER_INPROGRESS and has to retry the list_dkey RPC again and again.
Further more, if there are more dkeys created during the retry,
related list_dkey RPC may be hung for very long time.

To avoid such trouble, we may have to synchronously commit the DTX
that create the dkey.

Signed-off-by: Fan Yong <fan.yong@intel.com>

### CODE REVIEW

File: src/vos/vos_tree.c [hunk 1]
```diff
@@ -962,6 +962,34 @@ key_tree_prepare(struct vos_object *obj, daos_epoch_t epoch,
...
+			 * redundancy group.
+			 *
+			 * For VOS_ITER_DKEY operation, if some new dkeys which
+			 * DTXs are committable but related leader replicas are
+			 * on different targets, then in spite of list dkeys on
+			 * which replica, the client will get -DER_INPROGRESS
+			 * and has to retry the list_dkey RPC again and again.
+			 * Further more, if there are more dkeys created during
+			 * the retry, related list_dkey RPC may be hung there
+			 * for very long time.
```
Comment (NiuYawei) [inline src/vos/vos_tree.c]: So enumeration should just ignore the committables and return committed dkeys only, right?

File: src/vos/vos_tree.c [hunk 2]
```diff
@@ -962,6 +962,34 @@ key_tree_prepare(struct vos_object *obj, daos_epoch_t epoch,
...
+			 *
+			 * To avoid such trouble, we may have to synchronously
+			 * commit the DTX that create the dkey.
+			 */
```
Comment (NiuYawei) [inline src/vos/vos_tree.c]: Even if it's synchronous, -DER_INPROGRESS could be returned to client, and above issues are still existing?

I also have little concern on the DAOS array write performance where lots of dkeys could be generated on extending write.

File: src/vos/vos_tree.c (hunk 1 above)
Comment (Nasf-Fan) [reply src/vos/vos_tree.c]: No, the committable dkeys cannot be ignored, they needs to be returned for list_dkey RPC. But if the committable dkeys are created after the list_dkey RPC, in theory, they can be ignored. But from the implementation view, we have to distinguish whether the dkey is created before or after the list_dkey RPC, that will complex the logic.

File: src/vos/vos_tree.c [hunk 3]
```diff
@@ -962,6 +962,34 @@ key_tree_prepare(struct vos_object *obj, daos_epoch_t epoch,
...
+			 * the same redundancy group (for the same replicated
+			 * object), the leaders replica for these different
+			 * dkeys may reside on different targets in the same
+			 * redundancy group.
+			 *
+			 * For VOS_ITER_DKEY operation, if some new dkeys which
+			 * DTXs are committable but related leader replicas are
+			 * on different targets, then in spite of list dkeys on
+			 * which replica, the client will get -DER_INPROGRESS
+			 * and has to retry the list_dkey RPC again and again.
```
Comment (liuxuezhao) [inline src/vos/vos_tree.c]: just to confirm that it possible retry list_dkey RPC for some times, but should not be forever right?

is it possible for another options that don't do synchronize commit when key create, but do two other changes:
1) when server serves list_dkey, for -DER_INPROGRESS case it also returns one (the first) conflicted DTX
2) when client retry list_dkey RPC to leader, it carries a flag ("FORCE_COMMIT" or sth else) and the collected conflicted DTX array, when server do synchronous commit for those DTXs and reply to client;
3) based on 2's reply, client maybe can get its result or retry...

seems it is similar as EC fetch's EINPROGRESS handling

**Comment by Nasf-Fan** [comment]
Need more consideration, close it.

**Pull request closed without merging**